nirvana.go

```diff
@@ -17,54 +17,247 @@ limitations under the License.
+			service.ParseRequestForm(),
+		),
+		Modifier(
+			service.FirstContextParameter(),
+			service.ConsumeAllIfComsumesIsEmpty(),
+			service.ProduceAllIfProducesIsEmpty(),
+			service.ConsumeNoneForHTTPGet(),
+			service.ConsumeNoneForHTTPDelete(),
+			service.ProduceNoneForHTTPDelete(),
+		),
+	)
+}
+
+// NewConfig creates a pure config.
+func NewConfig() *Config {
+	return &Config{
+		IP:          "",
+		Port:        80,
+		Logger:      &log.SilentLogger{},
+		Filters:     []service.Filter{},
+		Descriptors: []definition.Descriptor{},
+		Modifiers:   []service.DefinitionModifier{},
+		configSet:   make(map[string]interface{}),
+	}
+}
+
+// server is nirvana server.
+type server struct {
+	config *Config
+}
+
+// NewServer creates a nirvana server.
+func NewServer(c *Config) Server {
+	return &server{
+		config: c,
+	}
+}
+
+var noConfigInstaller = errors.InternalServerError.Build("Nirvana:NoConfigInstaller", "no config installer for external config name ${name}")
+
+func (s *server) builder() (service.Builder, error) {
+	builder := service.NewBuilder()
+	builder.SetLogger(s.config.Logger)
+	builder.AddFilter(s.config.Filters...)
+	builder.SetModifier(s.config.Modifiers.Combine())
+	if err := builder.AddDescriptor(s.config.Descriptors...); err != nil {
+		return nil, err
+	}
+	err := s.config.forEach(func(name string, config interface{}) error {
+		installer := ConfigInstallerFor(name)
+		if installer == nil {
+			return noConfigInstaller.Error(name)
 		}
+		err := installer.Install(builder, s.config)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return builder, nil
+}
+
+// Serve starts to listen and serve requests.
+// The method won't return except an error occurs.
+func (s *server) Serve() error {
+	builder, err := s.builder()
+	if err != nil {
+		return err
 	}
-	if c.EnableMetrics {
-		metrics.Metrics{}.Install(s, c.MetricsPath)
+	service, err := builder.Build()
+	if err != nil {
+		return err
+	}
+	server := &http.Server{
+		Addr:    fmt.Sprintf("%s:%d", s.config.IP, s.config.Port),
+		Handler: service,
+	}
+	return server.ListenAndServe()
+}
+
+// ConfigInstaller is used to install config to service builder.
+type ConfigInstaller interface {
+	// Name is the external config name.
+	Name() string
+	// Install installs stuffs before server starting.
+	Install(builder service.Builder, config *Config) error
+	// Uninstall uninstalls stuffs after server terminating.
+	Uninstall(builder service.Builder, config *Config) error
+}
+
+var installers = map[string]ConfigInstaller{}
+
+// ConfigInstallerFor gets installer by name.
+func ConfigInstallerFor(name string) ConfigInstaller {
+	return installers[name]
+}
+
+// RegisterConfigInstaller registers a config installer.
+func RegisterConfigInstaller(ci ConfigInstaller) {
+	if ConfigInstallerFor(ci.Name()) != nil {
+		panic(fmt.Sprintf("Config installer %s has been installed.", ci.Name()))
+	}
+	installers[ci.Name()] = ci
+}
+
+// IP sets ip into config.
```

Comment:
Following function comment is incorrect, you should say:

```
IP returns a Configurer to set ip into config.
```

Same for others, e.g.

```
Descriptor returns a Configurer to add descriptors into config.
```